drivers/led/led_gpio.c

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/gpio.h>
+#include <drivers/led.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(led_gpio, CONFIG_LED_GPIO_LOG_LEVEL);
+
+struct gpio_cfg {
+	const char *port;
+	u8_t pin;
+	u8_t polarity;
+};
+
+struct led_dts_cfg {
+	struct gpio_cfg gpio;
+	const char *label;
+	u32_t id;
+};
+
+struct led_gpio_data {
+	struct device *gpio_dev;
+	const struct led_dts_cfg *cfg;
+	bool state;
+};
+
+struct led_gpio_cfg {
+	const struct led_dts_cfg *led;
+	u32_t led_cnt;
+};
+
+static inline struct led_gpio_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline const struct led_gpio_cfg *get_dev_config(struct device *dev)
+{
+	return dev->config->config_info;
+}
+
+static struct led_gpio_data *led_parameters_get(struct device *dev, u32_t led)
+{
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		if (data[i].cfg->id == led) {
+			return &data[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int gpio_led_set_no_delay(struct led_gpio_data *data, bool state)
+{
+	int err;
+	u32_t pin_value;
+
+	data->state = state;
+
+	if (data->cfg->gpio.polarity & GPIO_INT_ACTIVE_HIGH) {
+		pin_value = (state ? 1 : 0);
+	} else {
+		pin_value = (state ? 0 : 1);
+	}
+
+	err = gpio_pin_write(data->gpio_dev, data->cfg->gpio.pin,
+			     pin_value);
+	if (err) {
+		LOG_DBG("%s turn %s error %d",
+			data->cfg->label,
+			state ? "on" : "off",
+			err);
+	} else {
+		LOG_DBG("%s turn %s", data->cfg->label,
+			state ? "on" : "off");
+	}
+
+	return err;
+}
+
+static int gpio_led_set(struct device *dev, u32_t index, bool state)
+{
+	return gpio_led_set_no_delay(data, state);
+}
+
+static int gpio_led_on(struct device *dev, u32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, true);
+}
+
+static int gpio_led_off(struct device *dev, u32_t led)
+{
+	struct led_gpio_data *data = led_parameters_get(dev, led);
+
+	if (!data) {
+		return -EINVAL;
+	}
+
+	return gpio_led_set(data, false);
+}
+
+static const struct led_driver_api led_api = {
+	.on = gpio_led_on,
+	.off = gpio_led_off,
+};
+
+static int gpio_led_init(struct device *dev)
+{
+	int err;
+	const struct led_gpio_cfg *cfg = get_dev_config(dev);
+	struct led_gpio_data *data = get_dev_data(dev);
+	u32_t pin_value;
+
+	for (size_t i = 0; i < cfg->led_cnt; i++) {
+		data[i].gpio_dev = device_get_binding(cfg->led[i].gpio.port);
+		if (!data[i].gpio_dev) {
+			LOG_DBG("Failed to get GPIO device for port %s pin %d",
+				cfg->led[i].gpio.port, cfg->led[i].gpio.pin);
+			return -EINVAL;
+		}
+
+		/* Set all leds pin as output */
+		err = gpio_pin_configure(data[i].gpio_dev, cfg->led[i].gpio.pin,
+					 GPIO_DIR_OUT);
```

Comment:
Hi @KAGA164 ,

I think you can pass the GPIO_INT_ACTIVE_{LOW,HIGH} flag to gpio_pin_configure(). This way you can call gpio_pin_set() directly with the logical value. This will handle the polarity directly. Maybe you also need to pass all the other GPIO flags found in the gpio DT property.